src/lib/core/utils/data-element-constraints.ts

```diff
@@ -319,3 +354,38 @@ export function mergeMaxNumValues(
   );
   return mergedMaxNumValues === Infinity ? undefined : mergedMaxNumValues;
 }
+
+/**
+ * returns an array of entities that are either the same entity as the provided variable, or its ancestors
+ */
+function ancestorEntitiesForVariable(
+  variable: VariableDescriptor,
+  entities: StudyEntity[]
+): StudyEntity[] {
+  const ancestors = entities.reduceRight((ancestors, entity) => {
+    if (
+      entity.id === variable.entityId ||
+      entity.children?.includes(ancestors[0])
+    ) {
+      ancestors.unshift(entity);
+    }
+    return ancestors;
+  }, [] as StudyEntity[]);
+
+  return ancestors;
+}
```

Comment:
There is an assumption that `entities` are in a `preorder` traversal of the tree. That might be a risky assumption to make. At the very least, we should add a comment.